evm/src/contracts/CircleIntegration/Governance.sol

```diff
@@ -0,0 +1,247 @@
+
+import {State} from "./State.sol";
+import {
+    SequentialRegistration,
+    getRegisteredEmitters,
+    getChainToDomain,
+    getConsumedVaas,
+    getDomainToChain,
+    getSequentialRegistration
+} from "./Storage.sol";
+
+abstract contract Governance is IGovernance, State, ERC1967Upgrade {
+    using BytesParsing for bytes;
+
+    error UnsupportedGovernanceAction();
+
+    uint16 constant GOVERNANCE_CHAIN = 1;
+    bytes32 constant GOVERNANCE_EMITTER =
+        0x0000000000000000000000000000000000000000000000000000000000000004;
+
+    // "CircleIntegration" (left-zero-padded)
+    bytes32 constant GOVERNANCE_MODULE =
+        0x000000000000000000000000000000436972636c65496e746567726174696f6e;
+
+    uint8 constant UPDATE_WORMHOLE_FINALITY = 1;
+    uint8 constant REGISTER_EMITTER_AND_DOMAIN = 2;
+    uint8 constant CONTRACT_UPGRADE = 3;
+
+    /**
+     * @notice `updateWormholeFinality` changes the wormhole message consistencyLevel.
+     * @param encodedVaa Attested Wormhole governance message with the following
+     * relevant fields:
+     * - Field Bytes Type Index
+     * - finality 1 uint8 35
+     */
+    function updateWormholeFinality(bytes memory encodedVaa) public pure {
+        // This is here to silence an unused variable warning.
+        encodedVaa;
+
+        revert UnsupportedGovernanceAction();
+    }
+
+    /// @inheritdoc IGovernance
+    function registerEmitterAndDomain(bytes memory encodedVaa) public {
+        (IWormhole.VM memory vaa, uint256 offset) =
+            _verifyAndConsumeGovernanceMessage(encodedVaa, REGISTER_EMITTER_AND_DOMAIN);
+
+        // Registering emitters should only be relevant for this contract's chain ID,
+        // unless the target chain is 0 (which means all chains).
+        uint16 targetChain;
+        (targetChain, offset) = vaa.payload.asUint16Unchecked(offset);
+
+        // NOTE: Reverting with Error(string) comes from the old implementation, so we preserve it.
+        require(targetChain == 0 || targetChain == _chainId, "invalid target chain");
+
+        uint16 foreignChain;
+        (foreignChain, offset) = vaa.payload.asUint16Unchecked(offset);
+
+        // NOTE: Reverting with Error(string) comes from the old implementation, so we preserve it.
+        require(foreignChain != 0 && foreignChain != _chainId, "invalid chain");
+
+        mapping(uint16 => bytes32) storage registeredEmitters = getRegisteredEmitters();
+
+        // Ideally we would like to be able to update foreign emitters with governance VAAs instead
+        // of having to upgrade the contract to change this contract's state. But currently
+        // governance VAA sequence numbers and timestamps are not reliable enough to order these
+        // registrations. Until either value becomes reliable, we will store the timestamp in the
+        // sequential registration storage slot.
+        mapping(uint16 => SequentialRegistration) storage sequentialRegistration =
```

Comment:
Should we move this instantiation closer to where the mapping is used?

Reply:
Moved it to right before where it is used: https://github.com/wormhole-foundation/wormhole-circle-integration/pull/48/commits/90bcbce498ec49ffd5618f82000d8b1f00862b8e